File: python/memory_core.py
"""
memory_core.py - Long-term memory store (PRD: episodic + semantic, retrieval, forgetting).

Implements:
- Episodic: concrete events (e.g. "user asked X, we did Y").
- Semantic: distilled insights (e.g. "user prefers Chinese", "tool X needs param Y").
- Retrieval: score = α·Recency + β·Importance + γ·Relevance (keyword-based relevance).
- Forgetting: garbage_collect by TTL and max_items.

Storage: uses optional MemoryStore (from storage abstraction); when None, uses file (MEMORY_FILE).
"""

import json
import os
import re
import time
import uuid
from pathlib import Path
from typing import Any, Dict, List, Optional, Tuple, TYPE_CHECKING

from config import WORKDIR

if TYPE_CHECKING:
    from storage.protocols import MemoryStore

# Set AGENT_MEMORY=0 to disable long-term memory (retrieve returns [], consolidate no-op)
MEMORY_ENABLED = os.environ.get("AGENT_MEMORY", "1").strip() not in ("0", "false", "False")

# Default storage under workspace so it persists across runs (used when _memory_store is None)
MEMORY_DIR = Path(os.environ.get("AGENT_MEMORY_DIR", str(WORKDIR / "data" / "memory")))
MEMORY_FILE = MEMORY_DIR / "memories.json"
PREFERENCES_FILE = Path(os.environ.get("AGENT_PREFERENCES_FILE", str(MEMORY_DIR / "preferences.json")))
PATTERNS_FILE = Path(os.environ.get("AGENT_PATTERNS_FILE", str(MEMORY_DIR / "patterns.json")))

# Optional store injected by get_storage_backend(); when set, load/save use it instead of file
_memory_store: Optional["MemoryStore"] = None


def set_memory_store(store: Optional["MemoryStore"]) -> None:
    """Inject MemoryStore (e.g. from get_default_memory_store()). File used when None."""
    global _memory_store
    _memory_store = store

# Short-term: max conversation turns to keep in context (sliding window)
MAX_HISTORY_TURNS = 30

# Long-term caps
MAX_MEMORY_ITEMS = 500
MAX_MEMORY_AGE_DAYS = 90

# Retrieval weights (α + β + γ = 1)
ALPHA_RECENCY = 0.25
BETA_IMPORTANCE = 0.35
GAMMA_RELEVANCE = 0.40


def _ensure_dir() -> None:
    MEMORY_DIR.mkdir(parents=True, exist_ok=True)


def _load_memories() -> List[Dict[str, Any]]:
    if _memory_store is not None:
        return _memory_store.load()
    _ensure_dir()
    if not MEMORY_FILE.exists():
        return []
    try:
        with open(MEMORY_FILE, "r", encoding="utf-8") as f:
            return json.load(f)
    except (json.JSONDecodeError, OSError):
        return []


def _save_memories(items: List[Dict[str, Any]]) -> None:
    sanitized = sanitize_for_storage(items)
    if _memory_store is not None:
        _memory_store.save(sanitized)
        return
    _ensure_dir()
    with open(MEMORY_FILE, "w", encoding="utf-8") as f:
        json.dump(sanitized, f, ensure_ascii=False, indent=2)


# -----------------------------------------------------------------------------
# Privacy: sensitive data detection and sanitization before storage
# -----------------------------------------------------------------------------

SENSITIVE_PATTERNS = [
    (re.compile(r"\b(api[_-]?key|apikey)\s*[:=]\s*[\w\-]+", re.I), "***"),
    (re.compile(r"\b(password|passwd|pwd)\s*[:=]\s*\S+", re.I), "***"),
    (re.compile(r"\b(token|secret|auth)\s*[:=]\s*[\w\-\.]+", re.I), "***"),
    (re.compile(r"\b(credit[_\s]?card|card\s*#?)\s*[:=]?\s*\d[\d\s\-]+", re.I), "***"),
]


def detect_sensitive_data(text: str) -> bool:
    """True if text appears to contain sensitive information."""
    if not text or not isinstance(text, str):
        return False
    lower = text.lower()
    return any(
        lower.find(p) >= 0
        for p in ("api_key", "api-key", "password", "token", "secret", "credential", "private_key")
    )


def sanitize_for_storage(data: Any) -> Any:
    """
    Return a deep copy of data with sensitive string values redacted.
    Used before persisting memories or session data.
    """
    if isinstance(data, dict):
        return {k: sanitize_for_storage(v) for k, v in data.items()}
    if isinstance(data, list):
        return [sanitize_for_storage(v) for v in data]
    if isinstance(data, str):
        out = data
        for pat, repl in SENSITIVE_PATTERNS:
            out = pat.sub(repl, out)
        return out
    return data


# -----------------------------------------------------------------------------
# Long-term memory: user preferences, project knowledge, learned patterns
# -----------------------------------------------------------------------------

def load_user_preferences() -> Dict[str, Any]:
    """
    Load persistent user preferences (LTM). Returns dict with model, permissions, etc.
    Defaults if file missing. Path: AGENT_PREFERENCES_FILE or data/memory/preferences.json.
    """
    path = PREFERENCES_FILE if PREFERENCES_FILE.is_absolute() else MEMORY_DIR / PREFERENCES_FILE.name
    if not path.exists():
        return {
            "model": os.environ.get("MODEL_ID", "gpt-4o"),
            "permissions": {},
            "features": [],
        }
    try:
        with open(path, "r", encoding="utf-8") as f:
            return json.load(f)
    except (json.JSONDecodeError, OSError):
        return {"model": os.environ.get("MODEL_ID", "gpt-4o"), "permissions": {}, "features": []}


def load_project_knowledge() -> Optional[str]:
    """
    Load project-specific knowledge from CLAUDE.md in workspace (LTM).
    Returns content or None if file does not exist.
    """
    claude_md = WORKDIR / "CLAUDE.md"
    if not claude_md.exists():
        return None
    try:
        return claude_md.read_text(encoding="utf-8").strip()
    except OSError:
        return None


def _load_patterns_raw() -> List[Dict[str, Any]]:
    """Load patterns list from file."""
    path = PATTERNS_FILE if PATTERNS_FILE.is_absolute() else MEMORY_DIR / PATTERNS_FILE.name
    if not path.exists():
        return []
    try:
        with open(path, "r", encoding="utf-8") as f:
            data = json.load(f)
        return data if isinstance(data, list) else []
    except (json.JSONDecodeError, OSError):
        return []


def _save_patterns_raw(items: List[Dict[str, Any]]) -> None:
    path = PATTERNS_FILE if PATTERNS_FILE.is_absolute() else MEMORY_DIR / PATTERNS_FILE.name
    MEMORY_DIR.mkdir(parents=True, exist_ok=True)
    with open(path, "w", encoding="utf-8") as f:
        json.dump(sanitize_for_storage(items), f, ensure_ascii=False, indent=2)


def record_pattern(pattern: Dict[str, Any]) -> None:
    """
    Record a learned pattern (LTM). E.g. successful tool use, workflow step.
    pattern should have: type, and optionally name, frequency, last_used, context.
    """
    now = time.time()
    patterns = _load_patterns_raw()
    name = pattern.get("name") or pattern.get("type", "")
    existing = next((p for p in patterns if p.get("name") == name or p.get("type") == pattern.get("type") and p.get("context") == pattern.get("context")), None)
    if existing:
        existing["frequency"] = existing.get("frequency", 0) + 1
        existing["last_used"] = now
        for k, v in pattern.items():
            if k not in ("frequency", "last_used"):
                existing[k] = v
    else:
        patterns.append({
            **pattern,
            "frequency": pattern.get("frequency", 1),
            "last_used": pattern.get("last_used", now),
        })
    # Cap total patterns
    patterns.sort(key=lambda p: (-p.get("frequency", 0), -p.get("last_used", 0)))
    _save_patterns_raw(patterns[:500])


def get_relevant_patterns(query: str, limit: int = 5) -> List[Dict[str, Any]]:
    """Return patterns whose context/name match query (simple keyword overlap). Used for LTM retrieval."""
    if not query or not query.strip():
        return []
    words = set(_normalize_text(query))
    patterns = _load_patterns_raw()
    scored = []
    for p in patterns:
        ctx = (p.get("context") or "") + " " + (p.get("name") or "") + " " + (p.get("type") or "")
        ctx_words = set(_normalize_text(ctx))
        overlap = len(words & ctx_words) / max(1, len(words))
        scored.append((overlap, p))
    scored.sort(key=lambda x: -x[0])
    return [p for _, p in scored[:limit]]


def _normalize_text(text: str) -> List[str]:
    """Simple tokenization for relevance: lowercase, split on non-alnum."""
    text = (text or "").strip().lower()
    return [w for w in re.split(r"\W+", text) if len(w) > 1]


def _relevance_score(query: str, content: str) -> float:
    """0..1 overlap: Jaccard-like over word sets."""
    q_words = set(_normalize_text(query))
    c_words = set(_normalize_text(content))
    if not q_words:
        return 0.0
    overlap = len(q_words & c_words) / len(q_words)
    return min(1.0, overlap + 0.1 * len(q_words & c_words) / max(1, len(c_words)))


def _recency_score(created_ts: float, now: float) -> float:
    """Exponential decay by days. Newer = higher."""
    days_ago = (now - created_ts) / 86400
    return 0.99 ** days_ago


def write_event(content: str, importance: float = 0.5) -> None:
    """Write an episodic memory (concrete event)."""
    items = _load_memories()
    now = time.time()
    items.append({
        "id": str(uuid.uuid4()),
        "type": "episodic",
        "content": (content or "").strip()[:2000],
        "importance": max(0.0, min(1.0, importance)),
        "created_at": now,
        "last_accessed": now,
    })
    _save_memories(items)


def write_insight(content: str, importance: float = 0.6) -> None:
    """Write a semantic memory (distilled fact / preference)."""
    items = _load_memories()
    now = time.time()
    items.append({
        "id": str(uuid.uuid4()),
        "type": "semantic",
        "content": (content or "").strip()[:2000],
        "importance": max(0.0, min(1.0, importance)),
        "created_at": now,
        "last_accessed": now,
    })
    _save_memories(items)


def retrieve(
    query: str,
    k: int = 5,
    alpha: float = ALPHA_RECENCY,
    beta: float = BETA_IMPORTANCE,
    gamma: float = GAMMA_RELEVANCE,
) -> List[Tuple[str, float, Dict[str, Any]]]:
    """
    Retrieve top-k memories by score = α·Recency + β·Importance + γ·Relevance.
    Returns list of (content, score, raw_record). If memory disabled, returns [].
    """
    if not MEMORY_ENABLED:
        return []
    items = _load_memories()
    if not items:
        return []
    now = time.time()
    scored: List[Tuple[float, Dict[str, Any]]] = []
    for m in items:
        rec = _recency_score(m["created_at"], now)
        imp = m.get("importance", 0.5)
        rel = _relevance_score(query, m.get("content", ""))
        score = alpha * rec + beta * imp + gamma * rel
        scored.append((score, m))
    scored.sort(key=lambda x: -x[0])
    return [(s[1]["content"], s[0], s[1]) for s in scored[:k]]


def garbage_collect(
    max_items: int = MAX_MEMORY_ITEMS,
    max_age_days: float = MAX_MEMORY_AGE_DAYS,
) -> int:
    """Remove oldest and low-importance items. Returns number removed."""
    items = _load_memories()
    now = time.time()
    cutoff_ts = now - max_age_days * 86400
    # Keep: recent or important; drop rest, then cap by max_items
    kept = [m for m in items if m["created_at"] >= cutoff_ts or m.get("importance", 0) >= 0.7]
    kept.sort(key=lambda m: (-m.get("importance", 0), -m["created_at"]))
    kept = kept[:max_items]
    removed = len(items) - len(kept)
    _save_memories(kept)
    return removed


def consolidate_turn(
    user_message: str,
    assistant_summary: str,
    tool_calls_summary: Optional[List[Dict[str, Any]]] = None,
) -> None:
    """
    After a turn: write an episodic memory of the exchange and optionally record
    learned patterns from tool use (LTM). No-op if memory is disabled.
    tool_calls_summary: optional list of {"name": str, "success": bool} for pattern recording.
    """
    if not MEMORY_ENABLED:
        return
    if not (user_message or assistant_summary):
        return
    # One-line episodic summary
    content = f"User: {user_message[:200]}. Assistant: {assistant_summary[:300]}."
    write_event(content, importance=0.4)
    # Record high-level turn pattern for LTM
    record_pattern({
        "type": "turn",
        "name": "conversation_turn",
        "context": user_message[:150],
    })
    if tool_calls_summary:
        for tc in tool_calls_summary:
            if tc.get("success"):
                record_pattern({
                    "type": "tool_use",
                    "name": tc.get("name", "tool"),
                    "context": user_message[:100],
                })
    # Lightweight garbage collect every 10 writes (could be done in background)
    items = _load_memories()
    if len(items) > MAX_MEMORY_ITEMS:
        garbage_collect(max_items=MAX_MEMORY_ITEMS, max_age_days=MAX_MEMORY_AGE_DAYS)

File: TypeScript/src/memoryCore.ts
/**
 * memoryCore.ts - Long-term memory store (episodic + semantic, retrieval, forgetting).
 *
 * Implements:
 * - Episodic: concrete events
 * - Semantic: distilled insights
 * - Retrieval: score = α·Recency + β·Importance + γ·Relevance
 * - Forgetting: garbage collection by TTL and max_items
 * - Sensitive data detection and sanitization
 */

import * as fs from 'fs';
import * as path from 'path';
import { v4 as uuidv4 } from 'uuid';
import { WORKDIR } from './config';

// ---------------------------------------------------------------------------
// MemoryStore interface (optional injection)
// ---------------------------------------------------------------------------

export interface MemoryStore {
    load(): Record<string, any>[];
    save(items: Record<string, any>[]): void;
}

let _memoryStore: MemoryStore | null = null;

export function setMemoryStore(store: MemoryStore | null): void {
    _memoryStore = store;
}

// Disable via AGENT_MEMORY=0
export const MEMORY_ENABLED =
    !['0', 'false'].includes((process.env.AGENT_MEMORY ?? '1').trim());

const MEMORY_DIR = process.env.AGENT_MEMORY_DIR ?? path.join(WORKDIR, 'data', 'memory');
const MEMORY_FILE = path.join(MEMORY_DIR, 'memories.json');
const PREFERENCES_FILE =
    process.env.AGENT_PREFERENCES_FILE ?? path.join(MEMORY_DIR, 'preferences.json');
const PATTERNS_FILE =
    process.env.AGENT_PATTERNS_FILE ?? path.join(MEMORY_DIR, 'patterns.json');

export const MAX_HISTORY_TURNS = 30;
export const MAX_MEMORY_ITEMS = 500;
export const MAX_MEMORY_AGE_DAYS = 90;

// Retrieval weights
export const ALPHA_RECENCY = 0.25;
export const BETA_IMPORTANCE = 0.35;
export const GAMMA_RELEVANCE = 0.40;

// ---------------------------------------------------------------------------
// File helpers
// ---------------------------------------------------------------------------

function ensureDir(dir: string): void {
    fs.mkdirSync(dir, { recursive: true });
}

function loadMemories(): Record<string, any>[] {
    if (_memoryStore) return _memoryStore.load();
    ensureDir(MEMORY_DIR);
    if (!fs.existsSync(MEMORY_FILE)) return [];
    try {
        return JSON.parse(fs.readFileSync(MEMORY_FILE, 'utf-8'));
    } catch {
        return [];
    }
}

function saveMemories(items: Record<string, any>[]): void {
    const sanitized = sanitizeForStorage(items);
    if (_memoryStore) {
        _memoryStore.save(sanitized);
        return;
    }
    ensureDir(MEMORY_DIR);
    fs.writeFileSync(MEMORY_FILE, JSON.stringify(sanitized, null, 2), 'utf-8');
}

// ---------------------------------------------------------------------------
// Privacy: sensitive data detection and sanitization
// ---------------------------------------------------------------------------

const SENSITIVE_PATTERNS: [RegExp, string][] = [
    [/\b(api[_-]?key|apikey)\s*[:=]\s*[\w\-]+/gi, '***'],
    [/\b(password|passwd|pwd)\s*[:=]\s*\S+/gi, '***'],
    [/\b(token|secret|auth)\s*[:=]\s*[\w\-.]+/gi, '***'],
    [/\b(credit[_\s]?card|card\s*#?)\s*[:=]?\s*\d[\d\s\-]+/gi, '***'],
];

export function detectSensitiveData(text: string): boolean {
    if (!text || typeof text !== 'string') return false;
    const lower = text.toLowerCase();
    return [
        'api_key', 'api-key', 'password', 'token', 'secret', 'credential', 'private_key',
    ].some((p) => lower.includes(p));
}

export function sanitizeForStorage(data: any): any {
    if (typeof data === 'object' && data !== null) {
        if (Array.isArray(data)) return data.map(sanitizeForStorage);
        const result: Record<string, any> = {};
        for (const [k, v] of Object.entries(data)) {
            result[k] = sanitizeForStorage(v);
        }
        return result;
    }
    if (typeof data === 'string') {
        let out = data;
        for (const [pat, repl] of SENSITIVE_PATTERNS) {
            out = out.replace(pat, repl);
        }
        return out;
    }
    return data;
}

// ---------------------------------------------------------------------------
// User preferences, project knowledge, patterns
// ---------------------------------------------------------------------------

export function loadUserPreferences(): Record<string, any> {
    const prefPath = path.isAbsolute(PREFERENCES_FILE)
        ? PREFERENCES_FILE
        : path.join(MEMORY_DIR, path.basename(PREFERENCES_FILE));
    if (!fs.existsSync(prefPath)) {
        return {
            model: process.env.MODEL_ID ?? 'gpt-4o',
            permissions: {},
            features: [],
        };
    }
    try {
        return JSON.parse(fs.readFileSync(prefPath, 'utf-8'));
    } catch {
        return { model: process.env.MODEL_ID ?? 'gpt-4o', permissions: {}, features: [] };
    }
}

export function loadProjectKnowledge(): string | null {
    const claudeMd = path.join(WORKDIR, 'CLAUDE.md');
    if (!fs.existsSync(claudeMd)) return null;
    try {
        return fs.readFileSync(claudeMd, 'utf-8').trim();
    } catch {
        return null;
    }
}

function loadPatternsRaw(): Record<string, any>[] {
    const pPath = path.isAbsolute(PATTERNS_FILE)
        ? PATTERNS_FILE
        : path.join(MEMORY_DIR, path.basename(PATTERNS_FILE));
    if (!fs.existsSync(pPath)) return [];
    try {
        const data = JSON.parse(fs.readFileSync(pPath, 'utf-8'));
        return Array.isArray(data) ? data : [];
    } catch {
        return [];
    }
}

function savePatternsRaw(items: Record<string, any>[]): void {
    const pPath = path.isAbsolute(PATTERNS_FILE)
        ? PATTERNS_FILE
        : path.join(MEMORY_DIR, path.basename(PATTERNS_FILE));
    ensureDir(MEMORY_DIR);
    fs.writeFileSync(
        pPath,
        JSON.stringify(sanitizeForStorage(items), null, 2),
        'utf-8',
    );
}

export function recordPattern(pattern: Record<string, any>): void {
    const now = Date.now() / 1000;
    const patterns = loadPatternsRaw();
    const name = pattern.name ?? pattern.type ?? '';
    const existing = patterns.find(
        (p) =>
            p.name === name ||
            (p.type === pattern.type && p.context === pattern.context),
    );
    if (existing) {
        existing.frequency = (existing.frequency ?? 0) + 1;
        existing.last_used = now;
        for (const [k, v] of Object.entries(pattern)) {
            if (k !== 'frequency' && k !== 'last_used') existing[k] = v;
        }
    } else {
        patterns.push({
            ...pattern,
            frequency: pattern.frequency ?? 1,
            last_used: pattern.last_used ?? now,
        });
    }
    patterns.sort((a, b) => -(a.frequency ?? 0) + (b.frequency ?? 0) || -(a.last_used ?? 0) + (b.last_used ?? 0));
    savePatternsRaw(patterns.slice(0, 500));
}

export function getRelevantPatterns(
    query: string,
    limit: number = 5,
): Record<string, any>[] {
    if (!query?.trim()) return [];
    const words = new Set(normalizeText(query));
    const patterns = loadPatternsRaw();
    const scored: [number, Record<string, any>][] = [];
    for (const p of patterns) {
        const ctx =
            (p.context ?? '') + ' ' + (p.name ?? '') + ' ' + (p.type ?? '');
        const ctxWords = new Set(normalizeText(ctx));
        let overlap = 0;
        for (const w of words) if (ctxWords.has(w)) overlap++;
        scored.push([overlap / Math.max(1, words.size), p]);
    }
    scored.sort((a, b) => b[0] - a[0]);
    return scored.slice(0, limit).map(([, p]) => p);
}

// ---------------------------------------------------------------------------
// Text helpers
// ---------------------------------------------------------------------------

function normalizeText(text: string): string[] {
    return (text ?? '').trim().toLowerCase().match(/\w{2,}/g) ?? [];
}

function relevanceScore(query: string, content: string): number {
    const qWords = new Set(normalizeText(query));
    const cWords = new Set(normalizeText(content));
    if (qWords.size === 0) return 0;
    let overlap = 0;
    for (const w of qWords) if (cWords.has(w)) overlap++;
    return Math.min(1.0, overlap / qWords.size + (0.1 * overlap) / Math.max(1, cWords.size));
}

function recencyScore(createdTs: number, now: number): number {
    const daysAgo = (now - createdTs) / 86400;
    return Math.pow(0.99, daysAgo);
}

// ---------------------------------------------------------------------------
// Core memory operations
// ---------------------------------------------------------------------------

export function writeEvent(content: string, importance: number = 0.5): void {
    const items = loadMemories();
    const now = Date.now() / 1000;
    items.push({
        id: uuidv4(),
        type: 'episodic',
        content: (content || '').trim().slice(0, 2000),
        importance: Math.max(0, Math.min(1, importance)),
        created_at: now,
        last_accessed: now,
    });
    saveMemories(items);
}

export function writeInsight(content: string, importance: number = 0.6): void {
    const items = loadMemories();
    const now = Date.now() / 1000;
    items.push({
        id: uuidv4(),
        type: 'semantic',
        content: (content || '').trim().slice(0, 2000),
        importance: Math.max(0, Math.min(1, importance)),
        created_at: now,
        last_accessed: now,
    });
    saveMemories(items);
}

export function retrieve(
    query: string,
    k: number = 5,
    alpha: number = ALPHA_RECENCY,
    beta: number = BETA_IMPORTANCE,
    gamma: number = GAMMA_RELEVANCE,
): [string, number, Record<string, any>][] {
    if (!MEMORY_ENABLED) return [];
    const items = loadMemories();
    if (!items.length) return [];
    const now = Date.now() / 1000;
    const scored: [number, Record<string, any>][] = [];
    for (const m of items) {
        const rec = recencyScore(m.created_at, now);
        const imp = m.importance ?? 0.5;
        const rel = relevanceScore(query, m.content ?? '');
        const score = alpha * rec + beta * imp + gamma * rel;
        scored.push([score, m]);
    }
    scored.sort((a, b) => b[0] - a[0]);
    return scored.slice(0, k).map(([score, m]) => [m.content, score, m]);
}

export function garbageCollect(
    maxItems: number = MAX_MEMORY_ITEMS,
    maxAgeDays: number = MAX_MEMORY_AGE_DAYS,
): number {
    const items = loadMemories();
    const now = Date.now() / 1000;
    const cutoffTs = now - maxAgeDays * 86400;
    let kept = items.filter(
        (m) => m.created_at >= cutoffTs || (m.importance ?? 0) >= 0.7,
    );
    kept.sort(
        (a, b) => -(a.importance ?? 0) + (b.importance ?? 0) || -a.created_at + b.created_at,
    );
    kept = kept.slice(0, maxItems);
    const removed = items.length - kept.length;
    saveMemories(kept);
    return removed;
}

export function consolidateTurn(
    userMessage: string,
    assistantSummary: string,
    toolCallsSummary?: { name: string; success: boolean }[],
): void {
    if (!MEMORY_ENABLED) return;
    if (!userMessage && !assistantSummary) return;
    const content = `User: ${userMessage.slice(0, 200)}. Assistant: ${assistantSummary.slice(0, 300)}.`;
    writeEvent(content, 0.4);
    recordPattern({
        type: 'turn',
        name: 'conversation_turn',
        context: userMessage.slice(0, 150),
    });
    if (toolCallsSummary) {
        for (const tc of toolCallsSummary) {
            if (tc.success) {
                recordPattern({
                    type: 'tool_use',
                    name: tc.name ?? 'tool',
                    context: userMessage.slice(0, 100),
                });
            }
        }
    }
    const items = loadMemories();
    if (items.length > MAX_MEMORY_ITEMS) {
        garbageCollect(MAX_MEMORY_ITEMS, MAX_MEMORY_AGE_DAYS);
    }
}
